libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc

```diff
@@ -0,0 +1,528 @@
+/** @file
+ *****************************************************************************
+
+ Implementation of interfaces for top-level Anemoi hash function gadgets.
+
+ See anemoi_gadget.hpp .
+
+ *****************************************************************************
+ * @author     This file is part of libsnark, developed by SCIPR Lab
+ *             and contributors (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+
+#ifndef LIBSNARK_GADGETLIB1_GADGETS_HASHES_ANEMOI_COMPONENTS_TCC_
+#define LIBSNARK_GADGETLIB1_GADGETS_HASHES_ANEMOI_COMPONENTS_TCC_
+
+namespace libsnark
+{
+
+// A R1CS constraint is a formal expression of the form
+//
+//                < A , X > * < B , X > = < C , X > ,
+//
+// where X = (x_0,x_1,...,x_m) is a vector of formal variables and
+// A,B,C each consist of 1+m elements in <FieldT> and <A, X> = \sum_i
+// (a_i x_i) is the dot product between vectors A and X. Equivalently,
+// the vectors A,B,C are linear combinations of X.
+//
+// A R1CS constraint is used to construct a R1CS constraint system.
+//
+// See also class \ef r1cs_constraint
+
+// R1CS constraints for the operation y = const_a x^2 + const_b with x = input,
+// y = output. This operation is realized by the components \ref
+// flystel_Q_gamma_prime_field_gadget and \ref
+// flystel_Q_delta_prime_field_gadget
+//
+// The operation is represented with one multiplication as
+//
+// (const_a x) * x = y-const_b
+//
+// for the variables vector X = (x0=1, x1=x, x2=y). This computation
+// is represented with 1 R1CS constraint:
+//
+// < A , X > * < B , X > = < C , X >
+//
+// where A =(0, const_a, 0), B=(0, 1, 0) and C =(-const_b, 0, 1)
+
+template<typename FieldT, size_t generator>
+flystel_Q_gamma_prime_field_gadget<FieldT, generator>::
+    flystel_Q_gamma_prime_field_gadget(
+        protoboard<FieldT> &pb,
+        const linear_combination<FieldT> &input,
+        const pb_variable<FieldT> &output,
+        const std::string &annotation_prefix)
+    : gadget<FieldT>(pb, annotation_prefix)
+#ifdef FLYSTEL_DEBUG
+    , beta(DEBUG_FLYSTEL_BETA)
+    , gamma(DEBUG_FLYSTEL_GAMMA)
+#else
+    , beta(FieldT(generator))
+    , gamma(FieldT(0))
+#endif // #ifdef FLYSTEL_DEBUG
+    , input(input)
+    , output(output)
+{
+    printf("[%s:%d] %s() input ", __FILE__, __LINE__, __FUNCTION__);
```

Comment:
All printfs need to be removed at some point.  Ideally we use test code to check specific values.

Reply:
Addressed.